Live attach/detach and its unit tests (#53)
ROCm/rocm-systems: projects/rocprofiler-compute/docs/how-to/live_attach_detach.rst

> .. meta::

Reviewer: We should ask Pratik to review this.

Reviewer 2: already updated with Pratik

   :description: ROCm Compute Profiler: using Live Attach Detach
   :keywords: ROCm Compute Profiler, Attach Detach

********************************************
Using Live Attach Detach in ROCm Compute Profiler
********************************************

> Live Attach/Detach is a new way for rocprofiler-compute of coupling with the workload process, which does not control the start and end of the workload application. When it runs, the application is already running. What the profiler does is only to attach to the workload process, collect the needed counter and detach, without changing the lifecycle of the workload.

Reviewer: Live Attach/Detach is a new feature of `rocprofiler-compute` that allows coupling with a workload process, without controlling its start or end. Instead, the application is already running when the profiler is invoked. The profiler simply attaches to the process, collects the required counters, and then detaches—without altering the lifecycle of the workload.

> Due to a specific attach is not repeatable, it can only collect a set of counter that the hardware is capable of doing during one run. Therefore, in current implementation, the user needs to specific the subset of goupe of counter and ensure that it's able to acquire during only one run. It can be achieved either by using "--block" parameter, for example, "--block 3.1.1 4.1.1 5.1.1", or by using single path argument such as "--set launch_stats".

Reviewer: Since a specific attach is not repeatable, it can only collect the set of counters that the hardware is capable of capturing in a single run. Therefore, in the current implementation, the user must specify a subset of counter groups that can be collected within one run. This can be done either by using the `--block` (for example, `--block 3.1.1 4.1.1 5.1.1`), or by providing a predefined set through the use of single pass counter collection `--set`.

Reviewer: > Since a specific attach is not repeatable, it can only collect the set of counters that the hardware is capable of capturing in a single run. Therefore, in the current implementation, the user must specify a subset of counter groups that can be collected within one run. This can be done either by using the `--block` (for example, --block 3.1.1 4.1.1 5.1.1), or by providing a predefined set through the use of single pass counter collection `--set`.

Should add documentation references in links to options `block` and `set`.

> The method of detach can be achieve by two way: 1. by setting parameter "--attach-duration-msec" of a specific time in terms of milliseconds, the detach will happen after this preset time since the start of subprocess of rocprofiler. 2. By clicking "Enter" key after the successful attach.

Reviewer 3: By clicking "Enter" key after the successful attach within the same profiling session.

> The method of detach can be achieve by two way: 1. by setting parameter "--attach-duration-msec" of a specific time in terms of milliseconds, the detach will happen after this preset time since the start of subprocess of rocprofiler. 2. By clicking "Enter" key after the successful attach within the same profiling terminal session. When successful attach happens, a message indicating it can be seen in the terminal log of the workload process.

Reviewer: Detachment can be achieved in two ways:

1. By setting the `--attach-duration-msec` parameter to a specific duration (in milliseconds). In this case, detachment occurs automatically after the specified time has elapsed since the `rocprofiler` subprocess started.
2. By pressing the Enter key after a successful attach within the same profiling terminal session. Upon a successful attach, a confirmation message is displayed in the terminal log of the workload process.

---------------------
Profiling options
---------------------
For using profiling options for PC sampling the configuration needed are:

* ``--attach-pid``: Should be the process ID of the process of workload's application.
* ``--attach-duration-msec``: (Optional) The is for setting up the synchronised detach and it's optional. Its unit is in milliseconds. When setting up, the detach will happen adter this time since rocprof starts. For example, setting it to 6000 yields 1 mins.

**Sample command:**

.. code-block:: shell

>    $ rocprof-compute profile -n try_live_attach_detach -b 3.1.1 4.1.1 5.1.1 --no-roof -VVV ----attach-pid <process id of workload>

Reviewer: Should provide example output for sample command.

Reviewer 2: for example output, do you mean the list of output file? Or the analyze output?

Reviewer: example output of what users should expect when using option `--attach-pid`

   $ rocprof-compute profile -n try_live_attach_detach --set launch_stats --no-roof -VVV --pid <process id of workload>

   $ rocprof-compute profile -n try_live_attach_detach -b 3.1.1 4.1.1 5.1.1 --no-roof -VVV --pid <process id of workload> --attach-duration-msec <time before detach>

   $ rocprof-compute profile -n try_live_attach_detach --set launch_stats --no-roof -VVV --pid <process id of workload> --attach-duration-msec <time before detach>

-----------------------
Analysis options
-----------------------
The analyze options for attach/detach are completely compatible with the non-attach/detach option

.. note::

  * Live Attach Detach feature is currently in BETA version. To enable Live Attach Detach, you have to have proper verson of rocprofiler-sdk and rocprofiler-register.
  * To make Live Attach Detach work, you must use "--block" or single path to limit the number of counter input files to 1. This limitation will be release in later version with implementation such as Iteration Mutiplexing.

>   * Due to limitation of rocporfiler-sdk, the attach can now only happen before HSA initialization. This will be solved in later release of rocprofiler-sdk

Reviewer 3: "only happen before HSA initialization", this is not direct meaningful info for the end user. Generally,  the end user has not much knowledge when HSA initialization happens.

.. meta::
   :description: ROCm Compute Profiler: using Live Attach Detach
   :keywords: ROCm Compute Profiler, Attach Detach

> ********************************************

Reviewer 4: ```suggestion
***********************************************************
```

> Using Live Attach Detach in ROCm Compute Profiler

Reviewer 4: ```suggestion
Using Live Attach/Detach in ROCm Compute Profiler
```

> ********************************************

Reviewer 4: ```suggestion
***********************************************************
```

> Live Attach/Detach is a new feature of rocprofiler-compute that allows coupling with a workload process, without controlling its start or end. Instead, the application is already running when the profiler is invoked. The profiler simply attaches to the process, collects the required counters, and then detaches—without altering the lifecycle of the workload.

Reviewer 4: ```suggestion
Live Attach/Detach is a new feature of ROCm Compute Profiler that allows coupling with a workload process, without controlling its start or end. The application can already be running before the profiler application is invoked. The profiler simply attaches to the process, collects the required counters, and then detaches—without altering the lifecycle of the workload.
```

> Since a specific attach is not repeatable, it can only collect the set of counters that the hardware is capable of capturing in a single run. Therefore, in the current implementation, the user must specify a subset of counter groups that can be collected within one run. This can be done either by using the --block (for example, --block 3.1.1 4.1.1 5.1.1), or by providing a predefined set through the use of single pass counter collection --set.

Reviewer 4: ```suggestion
A specific attach is not repeatable, and it can only collect the set of counters that the hardware is capable of capturing in a single run. As such, in the current implementation, you must specify a subset of counter groups that can be collected within one run. This can be done either by using the ``--block`` option (for example, --block 3.1.1 4.1.1 5.1.1) or by providing a predefined set through the use of single pass counter collection ``--set``.
```

Detachment can be achieved in two ways:

> By setting the --attach-duration-msec parameter to a specific duration (in milliseconds). In this case, detachment occurs automatically after the specified time has elapsed since the rocprofiler subprocess started.

Reviewer 4: ```suggestion
a) By setting the ``--attach-duration-msec`` parameter to a specific duration (in milliseconds). In this case, the detachment occurs automatically after the specified time has elapsed since the ``rocprof`` subprocess started.
```

> By pressing the Enter key after a successful attach within the same profiling terminal session. Upon a successful attach, a confirmation message is displayed in the terminal log of the workload process.

Reviewer 4: ```suggestion
b) By pressing the Enter key after a successful attach within the same profiling terminal session. Upon a successful attach, a confirmation message is displayed in the terminal log of the workload process.
```

---------------------
Profiling options
---------------------
For using profiling options for PC sampling the configuration needed are:

> * ``--attach-pid``: Should be the process ID of the process of workload's application.

Reviewer 4: Should it be `--attach-id`? The changelog file mentions `--attach-id`. The changelog and the detailed document need to be synced.

Reviewer 2: sorry, wrote it wrong in the change log. Let me correct

> * ``--attach-duration-msec``: (Optional) The is for setting up the synchronised detach and it's optional. Its unit is in milliseconds. When setting up, the detach will happen adter this time since rocprof starts. For example, setting it to 60000 yields 1 mins.

Reviewer 4: ```suggestion
* ``--attach-duration-msec``: (Optional) This is for setting up the synchronized detach, and is optional. Its unit is in milliseconds. When setting up, the detach will happen after this time has elapsed since the ``rocprof`` subprocess started. For example, setting it to 60000 yields 1 minute.
```

**Sample command:**

.. code-block:: shell
   $ rocprof-compute profile -n try_live_attach_detach -b 3.1.1 4.1.1 5.1.1 --no-roof -VVV --attach-pid <process id of workload>

   $ rocprof-compute profile -n try_live_attach_detach --set launch_stats --no-roof -VVV --attach-pid <process id of workload>

   $ rocprof-compute profile -n try_live_attach_detach -b 3.1.1 4.1.1 5.1.1 --no-roof -VVV --attach-pid <process id of workload> --attach-duration-msec <time before detach>

   $ rocprof-compute profile -n try_live_attach_detach --set launch_stats --no-roof -VVV --attach-pid <process id of workload> --attach-duration-msec <time before detach>

-----------------------
Analysis options
-----------------------

> The analyze options for attach/detach are completely compatible with the non-attach/detach option

Reviewer 4: ```suggestion
The analyze options for attach/detach are completely compatible with the non-attach/detach option.
```

.. note::

>   * Live Attach Detach feature is currently in BETA version. To enable Live Attach Detach, you have to have proper verson of rocprofiler-sdk and rocprofiler-register.

Reviewer 4: ```suggestion
  * Live Attach Detach feature is currently in BETA version. To enable Live/Attach Detach, you need to have the correct supported proper version of ROCprofiler-SDK and rocprofiler-register.
```

>   * To make Live Attach Detach work, you must use "--block" or single path to limit the number of counter input files to 1. This limitation will be release in later version with implementation such as Iteration Mutiplexing.

Reviewer 4: ```suggestion
  * To make the Live Attach/Detach feature work, you must use "--block" or a single path to limit the number of counter input files to one. This limitation will be removed in a later version with implementations such as Iteration Multiplexing.
```

>   * Due to limitation of rocporfiler-sdk, the attach can now only happen before HSA initialization. HSA initialization happens before the execution of the first HIP kernel call. It only happens once to save all the kernels' function signature, like function name and other launch parameters. Attaching after this stage misses all crucial infomations of HIP kernel and make it impossible to store output. This limitation will be solved in later releases of rocprofiler-sdk.

Reviewer 4: ```suggestion
  * Due to the limitation of ROCprofiler-SDK, the attach can now only happen before Heterogeneous System Architecture (HSA) initialization. HSA initialization happens before the execution of the first HIP kernel call. It only happens once to save all the kernels' function signature, such as the function name and other launch parameters. Attaching after this stage misses all crucial information of the HIP kernel and makes it impossible to store the output. This limitation will be solved in later releases of ROCprofiler-SDK.
```
